### ISSUE TITLE
Derivatives (wrt parameters and input values)

### ISSUE THREAD

**Issue by st--** [trthatcher/MLKernels.jl#2]
Hi,

to be able to use this kernel module in my Gaussian Process regression code - instead of rolling my own covariance kernels, which do basically the same, just not as fancy as your code - I will need the derivatives of kernels, both with respect to the parameters (for optimizing log marginal likelihoods) and with respect to input values (for training from and prediction of derivatives of function values, not just the function values themselves). What are your thoughts (if any) on the interface for that ?

Have a look at https://github.com/st--/juliagp/blob/master/covariances.jl for how I've implemented it so far...

-ST

**Comment by trthatcher** [comment]
Hey 

Sorry... I've been busy with work and I'm not that familiar with Github, so this totally passed under the radar!

Ahhhhh, calculus! Noooo! I'm meltinggggg!

Kidding, I think this sounds like a great feature. 

I'm not currently familiar with Gaussian processes. I'm assuming your covariance functions just positive definite kernel functions? Let's say you have a kernel k(x,y;a) where x and y are vectors (data) and a is a kernel parameter. You'll need something along these lines:
- The first derivative of k wrt a (dk/da) which will be a function of x,y and will return a scalar
- The first partial derivative of k wrt x or y (dk/dx) which will be a function x and y and will return a vector 
- The second partial derivative of k wrt x or y (dk^2/dx^2) which will be a function x and y and will return a square matrix
- The mixed partial derivative of k wrt x and y which will be a function of x and y and will return a matrix

Do I have that correct for the most part? I played around with your code and it looks like you have the first two defined for vectors and the latter two defined for scalars.

As for implementation, I like what you've started. I'd use x and y rather than strictly relying on positional use of _x or x_.

kernel_function is kind of verbose. Let's say it's changed to 'kernel'

I'm thinking it would make sense to have a parameters() function for each kernel that will return an ordered list of symbols (for each parameter). Users can specify which parameter to take the derivative with respect to using the index or the parameter.

Something like dkernel_dp(K, :a, x, y) or dkernel_dp(K, 1, x, y)

Then the others (for input values) are like:
dkernel_dx (dkernel_dy)
d2kernel_d2x (d2kernel_d2y)
d2kernel_dxdy

What are your thoughts?

**Comment by st--** [comment]
Yes, pretty much spot on! Yes, Gaussian Processes generally need posdef kernels (you need to invert the covariance matrix). Except I don't think I actually need the second _non_-mixed derivatives.

In my implementation, each covariance kernel only acts on a single piece of data - so a scalar if it's a 1D problem, or a vector of e.g. the position if it's multi-dimensional. Then covmat() iterates over the different points to construct the covariance matrix. The second mixed derivative was for prototyping around on a 1D problem in an IJulia notebook, so it's not yet used in any of the code, which is why it might not be as fleshed out / not working for vectors yet...

Your suggestions for how to construct the interface seem very sensible!

The only thing slightly more complicated are composite kernels - if you have a standard kernel, then the symbolic names for parameters are unambiguous, but what about, e.g., `5*GaussianKernel()+3*GaussianKernel()`? how to access the different scaling factors, and the parameters of the two instances of the Gaussian kernel?

Regarding the implementation, the derivatives very often use what's effectively the original kernel - particularly anything with an exp() in it! And in Gaussian Processes you often want to construct a covariance matrix between two lists of positions (so in data storage terms two two-dimensional arrays), block-diagonal as follows: top-left kernel(x,y), top-right kernel_dx(x,y), bottom-left kernel_dy(x,y) [or vice versa, I never remember], bottom-right kernel_dxdy(x,y)... so it would be nice if we could get the code in such a way that it does the minimum amount of calculations needed!

**Comment by st--** [comment]
Thought on composite kernels: Just use dot notation? E.g. for my example the symbols would be :a1, :k1.eta, :a2, :k2.eta. For something more complicated you might end up with :k1.k2.k2.c ... but I don't think there's any way around that.

NB for back-end code it obviously makes it look a lot neater to use Unicode symbols as you did, but anything exposed to the user - as these symbols - shouldn't require anything beyond ASCII...

**Comment by trthatcher** [comment]
I'm definitely targeting speed with this package. Do you have a good source for a description on fitting a Gaussian process? If your positions are p dimensional and you have n observations in one set and m in another set, then you're looking at calculating a (p+1)(p+1) dimension matrix for each pair (mn times).

Do you calculate this all at once? Or do you work with one block at a time and move onto the next one? I would imagine the former is memory intensive.

The Unicode characters are pretty gimmicky to be honest. You bring up a good point, I'm going to change all the parameters in the types back to ASCII but leave the Unicode in the printed output such as the descriptions. That way, the symbols used for accessing parameters will be strictly ASCII. I'm in complete support of the dot notation.

My intention with the composite kernels was intended as more of an exploratory, "cool" feature to try out more exotic kernels, so I'm less concerned about their functionality/performance at this point in time. I will definitely try to accommodate them, but for now I think it's best to get something working for the regular kernels.

I'm going to fix up those kernels now and finish making some changes to the kernel matrix calculations.

**Comment by trthatcher** [comment]
Related question: right now I have 'kernel_function' and 'kernel_matrix'.

They both accept array arguments. However, 'kernel_function' assumes the arrays are vectors and 'kernel_matrix' assumes they are observation/data matrices and operates on the rows/columns to generate the matrix.

Do you think I should condense this to one function "kernel" and do multiple dispatch? IE if arrays, then return kernel matrix. If vectors, kernel function that returns a scalar. My hesitation would just be that these are used very differently, so I would want to keep them sufficiently separate. Further, if I rely on multiple dispatch, you HAVE to wrap any column or row from a matrix with vec(). 
IE kernel(vec(A[1,:]), vec(A[2,:])) versus kernel_function(A[1,:], A[2,:])

Is 'kernel_function' too long? Especially if we're going to have derivatives. 'dkernel_function_dx'. Any ideas? 'kf' and 'km'? 'kernel' and 'kernel_matrix'?

**Comment by st--** [comment]
The first two chapters of http://www.gaussianprocess.org/gpml/chapters/ are a pretty good explanation into Gaussian Processes for Regression (which is what I use them for). TBH, I've learnt the most by implementing them and playing around with how it works.

A lot of that did involve composite kernels, and on http://www-ai.cs.uni-dortmund.de/weblab/static/api_docs/pyGPs/demoMaunaLoa.html you can find a run-through in a python package of what seems to be the canonical 1D example for GP regression. I'm not too bothered about speed of composite kernels at this point, but I think it's worth making available all the _functionality_ of derivatives for them. (I don't think this is a very complicated issue.)

Generally you can split GP regression into "teaching" on points x (with observed function values and/or derivatives y, y') and "prediction" on points xs; the first involves the covariance matrix C(x,x) and calculating its inverse (or its Cholesky decomposition); the second involves the covariance matrix C(x,xs). As you need the inverse in principle you need the whole at once, and you're right, for many dimensions and/or large data sets that definitely becomes an issue (though there's various approaches to speed that up and reduce memory footprint, perhaps trading off accuracy - such as sparsification - but in the first instance I'd focus on a solid implementation of basic GP regression, because that's what I'm playing with. For the big datasets I'm using our group's fortran code anyway. I'm glad we've got a compute node with 1.5 TB of RAM, but even that can still be limiting...).

I think it'd be better to keep kernel[_function] and kernel_matrix separate: you might want to calculate the point kernel of two vectors, or the kernel matrix of lists of 1D points - in either case it'd be two one-dimensional array inputs, but it's entirely separate things, in the first case returning a scalar, and in the second case a matrix - so I think it's worth making the distinction clear.

I like kernel() and kernel_matrix(). They clearly say what they do without being overly verbose. dkernel_dx, dkernel_dp are also nicely self-explanatory. Whether you call it kernel_matrix or kernelmatrix (mildly in favour of the latter) is a purely cosmetic issue...

**Comment by trthatcher** [comment]
Thanks for the resource! I'll definitely be checking that out.

I took a look at your first cut of the kernel derivatives. Abstracting the kernel function away from the dot product has worked out nicely in calculating the derivatives. It should be easy to apply this strategy to the rest of the kernels that are functions of a norm. The scalar product kernels should also be simple. The separable kernel will be a little more interesting...

I didn't have much time today, so I only made some aesthetic changes. I renamed euclidean_distance to norm2 and kernelize_scalar to kernelize to shorten things up a bit. Then I renamed a few of the derivative functions for consistency with the naming scheme mentioned above. Any thoughts/objections?

**Comment by st--** [comment]
I had wondered about kernelize[_scalar] and what the name means - still not sure about it, maybe there's a better name for it, but in the end here it doesn't matter because this shouldn't be exposed to users; it's a purely internal method.

Agreed that euclidean_distance is a bit of a mouthful, and dnorm2_dx sure is nicer. However, 'norm2' sounds like it'd be the 2-norm [i.e. norm(x,2) == sqrt(dot(x,x))] rather than the square of the 2-norm [norm(x,2)^2 == dot(x,x), which is what we actually use(†)]. Would you be fine renaming it to sq[uare]norm? Then also the second derivative can be called d2kernelize_dsqnorm2, in line with the mathematical notation: ∂²f/∂x²...

An alternative would be to move to the actual norm rather than epsilon'epsilon for the argument to kernelize() for EuclideanDistanceKernels, but maybe best to leave it as is for now.

(†) To be honest I think this is thinking about premature optimisation; in the end it probably _isn't_ the bottleneck, but all the same I was wondering if it wouldn't be more efficient to do something like (ϵ=x-y;dot(ϵ,ϵ)) instead of norm(x-y)^2 ... on the other hand, Base.generic_vecnorm2() [in linalg/generic.jl] which is what gets called eventually seems to do some scaling to minimize floating point errors or somesuch. Hum.

**Comment by trthatcher** [comment]
I am in no way attached to the names/labels of anything. If you have alternatives, please do make a suggestion. 

Also, I have not used Leibniz notation in forever... I totally switched up the order in the differential... I thought it was ∂²x for some reason. 

I'll just give you a quick overview of how things were intended to work and the story behind kernelize. Scanning through a data matrix and computing the covariance function/kernels gets really slow. However, I saw that the majority of kernels that I had come across were either an element-wise function of the scalar product or squared euclidean distance. I could compute the gramian quickly using BLAS. Similarly, transforming the gramian so that each ij element represented the norm squared of the lag (or error) between two vectors xi and xj was straight forward. This form the gramian I named "lagged_gramian". Separable kernels are similar story where the transformation is element wise on the vectors themselves (not their dot product or euclidean distance).

Basically, if X is your vectors and G is your gramian, Gl is your lagged gramian, and f is your kernelize function (returns matrix) then:

Scalar Product K = f(G)
Euclidean Distance K = f(Gl)
Separable K = f(X)*f(X)'

Does that make sense? If you have a better descriptor for this function, let me know.

As for the dots/norm calculation. I haven't had a chance to look too deeply into them... up until now I was just concerned about calculating the kernel matrix for something else I was working on. Element wise, the gramian returned by BLAS is a squared norm or dot product which was why I initially designed it to take in a dot(ϵ,ϵ) or dot(x,y). I would prefer to continue to use something like ϵ=x-y;dot(ϵ,ϵ).

Long story short, I'd like to stick with squared norm but I'm indifferent on how it's defined/named.

function sqnorm(x,y) 
ϵ=x-y;
dot(ϵ,ϵ)
end

Does that sound good?

**Comment by st--** [comment]
It does, and thanks for the explanation! The one thing I'm slightly confused by is you saying that the kernelize() function returns a matrix, because that's the original kernelize_scalar, and it takes the epsilon'epsilon scalar and returns a _scalar_, not a matrix - or what did I get wrong there?

For teaching and predicting function _values_ with a Gaussian Process all I need is the kernelmatrix() as implemented already. For teaching or predicting function _derivatives_ what's also needed is the covariance matrix between function values and derivatives and/or between derivatives and derivatives.

If the input space is d-dimensional, and there are N points, then the total number of input values is N*d. kernelmatrix() returns an N by N matrix. I would also need (a) kernelmatrix_dx() which calculates dkernel_dx() for each combination of x and y, this is a vector of length d, and stitch them together into a N*d by N matrix and (b) kernelmatrix_dxdy() which calculates d2kernel_dxdy, a d by d matrix, and stitching those together results in the final N*d by N*d matrix. Thoughts on how to implement that in a fast way ?

If that helps I can just implement the naïve+slow implementation for reference...

**Comment by trthatcher** [comment]
OH, sorry! I wasn't clear. 

Mathematically, there's a matrix valued function f that transforms a gramian into a kernel matrix.  Really all it does it apply another function (lets call it h) to each of the elements of G or Gl. 

So K = f(G) = [h(g_11) h(g_12); h(g_21) h(g_22)] for a 2d gramian matrix where g_ij is the ith, jth element.

In my implementation:
- f == kernel_matrix (or kernel if it's just two vectors)
- h == kernelize

The above works for kernels that are a function of the dot product or squared norm. For separable kernels, the kernelize function is applied to the data matrix. Then the gramian of the 'kernelized' data matrix is calculated.

As for the derivatives, I was going to approach it the same way you were. Let's say you those derivative matrices are denoted dKij.

Do you need it in this form: [dK11, dK12, dK13, ... , dKnn]? IE a bunch of matrix blocks like this [image](http://upload.wikimedia.org/math/7/6/6/766a13ae7edda6f35dacdedf306ed714.png)

Because if not, I think it would be beneficial to lay these matrices out in contiguous blocks of memory. You'd just iterate through your observations and calculate each matrix.  Since you know the size in advance, you could just reinterpret the block as a certain size. Basically you would have a (d_d)_(N*N) matrix. Each column would correspond to an entire d2kernel_dxdy matrix.

That's just my first thought, though. 

Sorry I haven't been able to add much the last couple days - I'm busy with work stuff.

**Comment by unoe** [comment]
Hi both, do you think it's possible to get the derivatives that you require using automatic differentiation?
http://www.juliadiff.org/

**Comment by kskyten** [comment]
What is the current status of derivatives? I see that there is a derivatives branch that seems to implement the functionality as discussed above. However I get the following error when I tried to use it:

```julia
ERROR: LoadError: LoadError: UndefVarError: FloatingPoint not defined
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
 in macro expansion at ./REPL.jl:95 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
while loading .../.julia/v0.5/MLKernels/src/auxfunctions.jl, in expression starting on line 32
while loading .../.julia/v0.5/MLKernels/src/MLKernels.jl, in expression starting on line 57
```
The branch is also lagging the main branch by quite a few commits. What needs to be done to make it work?

**Comment by trthatcher** [comment]
Closed in favour of #46
